cli/src/metrics.rs

```diff
@@ -65,67 +48,117 @@ impl metrics::Recorder for Recorder {
     }
 
     fn register_counter(&self, key: &Key) -> Counter {
-        let read = self
-            .inner
-            .counters
-            .read()
-            .unwrap_or_else(PoisonError::into_inner);
-        if let Some(counter) = read.get(key) {
-            Counter::from_arc(counter.clone())
-        } else {
-            drop(read);
-            let mut write = self
-                .inner
-                .counters
-                .write()
-                .unwrap_or_else(PoisonError::into_inner);
-            if let Some(counter) = write.get(key) {
-                Counter::from_arc(counter.clone())
-            } else {
-                let counter = Arc::new(AtomicU64::new(0));
-                write.insert(key.clone(), counter.clone());
-                Counter::from_arc(counter)
-            }
+        self.registry
+            .get_or_create_counter(key, |c| c.clone().into())
+    }
+
+    fn register_gauge(&self, key: &Key) -> Gauge {
+        self.registry.get_or_create_gauge(key, |c| c.clone().into())
+    }
+
+    fn register_histogram(&self, key: &Key) -> metrics::Histogram {
+        self.registry
+            .get_or_create_histogram(key, |c| c.clone().into())
+    }
+}
+
+struct BucketStatistics {
+    min: f64,
+    max: f64,
+    avg: f64,
+    count: usize,
+}
+
+impl BucketStatistics {
+    fn new() -> BucketStatistics {
+        BucketStatistics {
+            min: f64::NAN,
+            max: f64::NAN,
```

Comment:
Any reason to use `NaN` rather than `f64::MIN` and `f64::MAX` respectively? The current code might be ok but I believe `NaN` makes min and max order-sensitive, which is a bit hard to reason about.

Reply:
`NaN` doesn't make them order-sensitive: "If one of the arguments is NaN, then the other argument is returned." That seems to be exactly the semantics we want. That's in fact different from the nightly `minimum` and `maximum` functions, which propagate NaNs. Also, I rather like this code returning Nan when no samples are present; even though that can't actually happen in our setup.